Redesign element reference page (#14821)
PrairieLearn/PrairieLearn: docs/elements/index.md

# Elements for `question.html`

PrairieLearn [questions](../question/overview.md) are built from `pl-*` tags inside [`question.html`](../question/template.md) ("elements"). These elements define the behavior of the question: what the student sees, what they input, and how it is graded. This page lists every element grouped by what you're trying to do. If you don't find the element / behavior you are looking for, consider using a [custom grade function](../question/server.md#step-5-grade) (recommended) or [building a custom element](../devElements.md).

!!! tip

>     Already know which element you want? Jump to the [alphabetical reference](#reference) below.

Reviewer: The navigation menu on the side also has the elements in alphabetical order, those are likely easier than going to the reference below.

Author: Keeping this for now as it contains more information than the sidebar.

Reviewer: My intention was not to remove the reference below, but to include it in the text as an option.

> Most questions only need a handful of elements: [`pl-question-panel`](pl-question-panel.md), [`pl-multiple-choice`](pl-multiple-choice.md), [`pl-number-input`](pl-number-input.md), [`pl-string-input`](pl-string-input.md), [`pl-code`](pl-code.md), [`pl-figure`](pl-figure.md), and [`pl-answer-panel`](pl-answer-panel.md).

Reviewer: I think this list is biased. These are by no means the most common elements. Here's a suggestion for a revision of this paragraph.

```suggestion
The elements you will choose will depend on the type and complexity of your question. A simple question may consist of a question panel prompt (using [`pl-question-panel`](pl-question-panel.md)) followed by a single [submission element](#submission-elements). For more complex prompts, you may need to include [display elements](#display-elements) such as figures, code and sectioning. To provide additional feedback to students and graders, [additional content may be displayed conditionally](#panels-visibility) in different panels and contexts.
```

Author: Replaced with your suggested wording (with anchor links updated for the renamed section). Fixed in 6d3eab6.

---
<sub>🤖 Comment by Claude Code</sub>

- [`pl-integer-input`](pl-integer-input.md): Fill in an integer value such as -71, 0, 5, 21, and so on.
- [`pl-units-input`](pl-units-input.md): Fill in a number with units such as "1.5 m", "14 ms", "6.3 ft", and so on.
- [`pl-matrix-component-input`](pl-matrix-component-input.md): Fill in a matrix using a grid with one input per cell.
- [`pl-matrix-input`](pl-matrix-input.md): Supply a matrix in a supported programming language format.

??? note "Decision flowchart"

    ![Decision flowchart for numeric input elements](./element-decision-number.d2){layout="elk" pad="0" scale="1"}

### Text & symbolic input

When the response is a string, an essay, or a math expression.

- [`pl-string-input`](pl-string-input.md): Fill in a string value such as `"Illinois"`, `"GATTACA"`, `"computer"`, and so on.
- [`pl-rich-text-editor`](pl-rich-text-editor.md): Provide an in-browser formattable text editor for open-ended responses and essays.
- [`pl-sketch`](pl-sketch.md): Sketch curves and other mathematical objects (e.g., points, asymptotes, polygons).
- [`pl-string-input`](pl-string-input.md): Fill in a **string** value such as `"Illinois"`, `"GATTACA"`, `"computer"`, and so on.

> - [`pl-symbolic-input`](pl-symbolic-input.md): Fill in a **symbolic** value such as `x^2`, `sin(z)`, `mc^2`, and so on.

Reviewer: As mentioned above, I think it makes sense to include `pl-file-editor` here as well, even if it's repeated.

Author: Added `pl-file-editor` to Text & symbolic input, cross-linked back to Code & files. Fixed in 6d3eab6.

---
<sub>🤖 Comment by Claude Code</sub>

> - [`pl-sketch`](pl-sketch.md): Auto-gradable sketches of curves and other mathematical objects (e.g., points, asymptotes, polygons).

Reviewer: Again, `pl-file-upload` could be listed here for completeness.

Author: Added `pl-file-upload` to Drawings & images, cross-linked back to Code & files. Fixed in 6d3eab6.

---
<sub>🤖 Comment by Claude Code</sub>

    ![Decision flowchart for drawing input elements](./element-decision-drawing.d2){layout="elk" pad="0" scale="1"}

### Code & files

When the response is code or a file, written in the browser or uploaded from disk.

- [`pl-file-upload`](pl-file-upload.md): Provide a submission area to obtain a file with a specific naming scheme.

> - [`pl-file-editor`](pl-file-editor.md): Provide an in-browser code editor for writing and submitting code.

Reviewer: As implied above, I think it makes sense to have a quick note here about workspaces, and point to the `pl-workspace` element (as well as the workspaces docs).

Author: Added a paragraph in Code & files pointing to the workspaces docs and `<pl-workspace>`. Fixed in 6d3eab6.

---
<sub>🤖 Comment by Claude Code</sub>

??? note "Decision flowchart"

    ![Decision flowchart for code and file input elements](./element-decision-code.d2){layout="elk" pad="0" scale="1"}

## Display elements

Display elements render content to students — figures, code, data, files, and layout wrappers — without capturing any submission.

### Media & figures

For showing images, plots, graphs, or downloadable/previewable files alongside the question.

- [`pl-figure`](pl-figure.md): Embed an image file in the question.
- [`pl-file-download`](pl-file-download.md): Enable file downloads for data-centric questions.

> - [`pl-file-preview`](pl-file-preview.md): Display a preview of submitted files.

Reviewer: I think pl-file-preview should be moved to the "External feedback" portion. Most of the other elements here mostly apply to the question panel, but pl-file-preview applies to the submission content.

Author: Moved `pl-file-preview` into the new 'Submission feedback' subsection. Fixed in 6d3eab6.

---
<sub>🤖 Comment by Claude Code</sub>

- [`pl-graph`](pl-graph.md): Display graphs using [GraphViz DOT notation](https://graphviz.org/doc/info/lang.html), an adjacency matrix, or a [`networkx`](https://networkx.org/) graph.

### Code, data & variables

For showing pre-formatted code snippets, data frames, matrices, or Python variable values.

- [`pl-code`](pl-code.md): Display code rendered with the appropriate syntax highlighting.
- [`pl-python-variable`](pl-python-variable.md): Display formatted output of Python variables.
- [`pl-dataframe`](pl-dataframe.md): Display DataFrames with various options.
- [`pl-matrix-latex`](pl-matrix-latex.md): Display matrices using appropriate LaTeX commands for use in a mathematical expression.
- [`pl-variable-output`](pl-variable-output.md): Display matrices in code form for supported programming languages.

> - [`pl-external-grader-variables`](pl-external-grader-variables.md): Display expected and given variables for [externally graded questions](../externalGrading.md).

Reviewer: This is only useful for Python grader.

```suggestion
- [`pl-external-grader-variables`](pl-external-grader-variables.md): Display expected and given variables for [questions using the Python grader](../python-grader/index.md).
```

### Layout & content wrappers

For wrapping, templating, or sanitizing content rather than rendering a specific data type.

- [`pl-card`](pl-card.md): Display content within a card-styled component.
- [`pl-template`](pl-template.md): Display content from mustache templates.
- [`pl-overlay`](pl-overlay.md): Layer elements on top of one another at specific positions — commonly used to place input fields over a [`pl-figure`](pl-figure.md) or [`pl-drawing`](../pl-drawing/index.md) for labeled-diagram questions.

> - [`pl-xss-safe`](pl-xss-safe.md): Remove potentially unsafe content from HTML code.

Reviewer: The main use-case here is to sanitize student-provided content. I think it should be moved to the "External feedback" section.

Author: Moved `pl-xss-safe` into the new 'Submission feedback' subsection. Fixed in 6d3eab6.

---
<sub>🤖 Comment by Claude Code</sub>

## Panels & visibility

> These elements control where and when content appears — for example, showing feedback only in the submission panel, hiding instructor-only feedback during manual grading.

Reviewer: I'm wondering if it may be better to rename this section, and make it focus on feedback (purpose) rather than technical functionality (implementation). These elements determine what is shown (or not) to students once they submit something, so the main purpose here is to make sure that the feedback provided to students is clear.

Author: Renamed the section to 'Question flow and feedback' and reframed the intro around what students see at each stage of the lifecycle. Also renamed the third subsection to 'Submission feedback'. Fixed in 6d3eab6.

---
<sub>🤖 Comment by Claude Code</sub>

### Standard question panels

The three primary content panels shown to students throughout the question lifecycle.

- [`pl-question-panel`](pl-question-panel.md): Content to only show in the question panel. The actual question prompt should be placed here.
- [`pl-submission-panel`](pl-submission-panel.md): Content to only show in the submission panel. Additional feedback should be placed here.
- [`pl-answer-panel`](pl-answer-panel.md): Content to only show in the answer panel. Additional details about the correct answer should be placed here.

??? note "What the panels look like"

    ![Question panels](../question/panels.png)

### Conditional visibility

For making content appear in only some panels, or only during specific grading workflows.

- [`pl-hide-in-panel`](pl-hide-in-panel.md): Hide content in one or more display panels.
- [`pl-hide-in-manual-grading`](pl-hide-in-manual-grading.md): Hide content in the manual grading page.
- [`pl-manual-grading-only`](pl-manual-grading-only.md): Show content only during manual grading.

### Hints & external feedback

> - [`pl-hidden-hints`](pl-hidden-hints.md): Reveal hints progressively as a student submits more on the current variant.

Reviewer: Nit: I'm wondering if this may be better suited to be listed together with the display elements section, as it is mainly targeted as part of the question prompt. I don't have a strong opinion on it, though.

Author: Agreed — `pl-hidden-hints` is really part of the question prompt with progressive disclosure rather than feedback on a submission. Moved it under 'Layout & content wrappers' in Display elements. Fixed in 6bd8a65.

---
<sub>🤖 Comment by Claude Code</sub>

| Element                                             | Use instead                                                                                                                                                        | Notes                                                       |
| --------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ----------------------------------------------------------- |
| [`pl-dropdown`](pl-dropdown.md)                     | [`pl-multiple-choice`](pl-multiple-choice.md) with `display="dropdown"`, or [`pl-matching`](pl-matching.md) for multiple linked dropdowns sharing the same options | —                                                           |
| [`pl-prairiedraw-figure`](pl-prairiedraw-figure.md) | [`pl-drawing`](../pl-drawing/index.md)                                                                                                                             | —                                                           |

> | [`pl-variable-score`](pl-variable-score.md)         | (no replacement needed)                                                                                                                                            | All `v3` submission elements include score display options. |

Reviewer: ```suggestion
| Element                                             | Use instead                                                                                                                                                        |
| --------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| [`pl-dropdown`](pl-dropdown.md)                     | [`pl-multiple-choice`](pl-multiple-choice.md) with `display="dropdown"`, or [`pl-matching`](pl-matching.md) for multiple linked dropdowns sharing the same options |
| [`pl-prairiedraw-figure`](pl-prairiedraw-figure.md) | [`pl-drawing`](../pl-drawing/index.md)                                                                                                                             |
| [`pl-variable-score`](pl-variable-score.md)         | Not needed, all submission elements include score display options. |
```
